File: src/studica_control/src/components/examples/cobra_example.py
#!/usr/bin/env python3
"""Cobra — subscribe to reflectance readings from all 4 channels.

the cobra is a 4-channel infrared reflectance sensor. each channel outputs
a voltage between 0 and vref — low over dark surfaces, high over light ones.
use the 4 channel readings together to follow a line or detect surface changes.

Run:  python3 cobra_example.py
Requires: studica_launch.py running, cobra enabled in params.yaml

Topic:   subscribes to 'cobra' (std_msgs/Float32MultiArray) — set cobra.<sensor>.topic: cobra in params.yaml
           data[0..3] = voltage in volts for channels 0-3
Service: 'cobra_cmd' (studica_control/SetData)
  Commands:
    get_raw <channel>     — raw adc value for one channel (0-3)
    get_voltage <channel> — voltage in volts for one channel (0-3)
"""
import rclpy
from rclpy.node import Node
from std_msgs.msg import Float32MultiArray
from studica_control.srv import SetData


class CobraExample(Node):
    def __init__(self):
        super().__init__('cobra_example')
        self.sub = self.create_subscription(
            Float32MultiArray, 'cobra', self.on_reading, 10)
        self.client = self.create_client(SetData, 'cobra_cmd')
        self.get_logger().info('Listening on cobra topic (4 channels)...')

    def on_reading(self, msg):
        channels = msg.data
        if len(channels) < 4:
            return
        self.get_logger().info(
            f'ch0: {channels[0]:.3f}V  ch1: {channels[1]:.3f}V  '
            f'ch2: {channels[2]:.3f}V  ch3: {channels[3]:.3f}V')

    def get_voltage(self, channel: int):
        """Request the voltage for a single channel via service."""
        if not self.client.wait_for_service(timeout_sec=2.0):
            self.get_logger().error('cobra_cmd service not available')
            return
        req = SetData.Request()
        req.params = f'get_voltage {channel}'
        future = self.client.call_async(req)
        rclpy.spin_until_future_complete(self, future)
        if future.result():
            self.get_logger().info(
                f'channel {channel} voltage: {future.result().message} V')


def main():
    rclpy.init()
    node = CobraExample()
    rclpy.spin(node)
    rclpy.shutdown()


if __name__ == '__main__':
    main()

File: src/studica_control/src/components/examples/cpp/cobra_example.cpp
/*
 * cobra_example.cpp
 *
 * Subscribe to all 4 Cobra reflectance channels and print them side by side.
 * Demonstrates the get_voltage service command for a single channel.
 *
 * Run:      ros2 run studica_control cobra_example
 * Requires: studica_launch.py running, cobra enabled in params.yaml
 *           sensors: ["cobra"]
 *
 * Topics subscribed: /cobra/ch_0 … /cobra/ch_3  (std_msgs/Float32)
 *   voltage reading per channel at 20 Hz
 * Service: /cobra/cobra_cmd  (studica_control/SetData)
 *   initparams.n_encoder — channel to query (0-3)
 *   Commands: get_raw, get_voltage
 */

#include <array>
#include <chrono>
#include <cmath>
#include <memory>
#include <string>

#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/float32.hpp"
#include "studica_control/srv/set_data.hpp"

using namespace std::chrono_literals;

static const std::string SENSOR = "cobra";

class CobraExample : public rclcpp::Node {
public:
    CobraExample() : Node("cobra_example"), ch_({NAN, NAN, NAN, NAN}) {
        for (int i = 0; i < 4; ++i) {
            ch_subs_[i] = create_subscription<std_msgs::msg::Float32>(
                "/" + SENSOR + "/ch_" + std::to_string(i), 10,
                [this, i](std_msgs::msg::Float32::SharedPtr msg) {
                    ch_[i] = msg->data;
                });
        }

        client_ = create_client<studica_control::srv::SetData>("/" + SENSOR + "/cobra_cmd");

        // print all 4 channels at 10 Hz
        print_timer_ = create_wall_timer(100ms, std::bind(&CobraExample::print_channels, this));

        // poll channel 0 voltage via service every 5 seconds
        svc_timer_ = create_wall_timer(5s, std::bind(&CobraExample::poll_service, this));

        RCLCPP_INFO(get_logger(), "cobra example ready (sensor: %s)", SENSOR.c_str());
    }

private:
    void print_channels() {
        RCLCPP_INFO(get_logger(),
            "ch0: %.3fV  ch1: %.3fV  ch2: %.3fV  ch3: %.3fV",
            ch_[0], ch_[1], ch_[2], ch_[3]);
    }

    void poll_service() {
        if (!client_->wait_for_service(1s)) {
            RCLCPP_WARN(get_logger(), "cobra_cmd service not available");
            return;
        }
        auto req = std::make_shared<studica_control::srv::SetData::Request>();
        req->params = "get_voltage";
        req->initparams.n_encoder = 0;  // query channel 0
        client_->async_send_request(req,
            [this](rclcpp::Client<studica_control::srv::SetData>::SharedFuture f) {
                RCLCPP_INFO(get_logger(), "service ch0 voltage: %s V",
                            f.get()->message.c_str());
            });
    }

    std::array<rclcpp::Subscription<std_msgs::msg::Float32>::SharedPtr, 4> ch_subs_;
    rclcpp::Client<studica_control::srv::SetData>::SharedPtr client_;
    rclcpp::TimerBase::SharedPtr print_timer_;
    rclcpp::TimerBase::SharedPtr svc_timer_;
    std::array<float, 4> ch_;
};

int main(int argc, char *argv[]) {
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<CobraExample>());
    rclcpp::shutdown();
    return 0;
}
